sol/main.c

```diff
@@ -0,0 +1,551 @@
+	return -1;
+}
+
+static int
+sol_stage2(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret = req_queue_init(sol_conf);
+	if (ret < 0)
+		goto cleanup;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return ret;
+}
+
+int
+run_sol(struct net_config *net_conf, struct sol_config *sol_conf)
+{
+	int ret;
+
+	if (net_conf == NULL || sol_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (!net_conf->back_iface_enabled) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: back interface is required\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->pri_req_max_len == 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"sol: priority queue max len must be greater than 0\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->enq_burst_size == 0 || sol_conf->deq_burst_size == 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: priority queue enqueue and dequeue sizes must both be greater than 0\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->deq_burst_size > sol_conf->pri_req_max_len ||
+			sol_conf->enq_burst_size > sol_conf->pri_req_max_len) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: request queue enqueue and dequeue sizes must be less than the max length of the request queue\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->req_bw_rate <= 0 || sol_conf->req_bw_rate >= 1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"sol: request queue bandwidth must be in range (0, 1), but it has been specified as %f\n",
+			sol_conf->req_bw_rate);
+		ret = -1;
+		goto out;
+	}
+
+	ret = init_mailbox("sol_reqs", 2 * sol_conf->pri_req_max_len,
+		sizeof(struct priority_req), sol_conf->lcore_id,
+		&sol_conf->mb);
+	if (ret < 0)
+		goto out;
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 0, 1, sol_stage1, sol_conf);
+	if (ret < 0)
+		goto mb;
+
+	ret = launch_at_stage2(sol_stage2, sol_conf);
+	if (ret < 0)
+		goto stage1;
+
+	ret = launch_at_stage3("sol", sol_proc, sol_conf, sol_conf->lcore_id);
+	if (ret < 0)
+		goto stage2;
+
+	sol_conf->net = net_conf;
+
+	ret = 0;
+	goto out;
+
+stage2:
+	pop_n_at_stage2(1);
+stage1:
+	pop_n_at_stage1(1);
+mb:
+	destroy_mailbox(&sol_conf->mb);
+out:
+	return ret;
+}
+
+/*
+ * There should be only one sol_config instance.
+ * Return an error if trying to allocate the second instance.
+ *
+ * Use rte_calloc() to zero-out the instance and initialize the
+ * request queue list to guarantee that cleanup_sol() won't fail
+ * during initialization.
+ */
+struct sol_config *
+alloc_sol_conf(void)
+{
+	struct sol_config *sol_conf;
+	static rte_atomic16_t num_sol_conf_alloc = RTE_ATOMIC16_INIT(0);
+	if (rte_atomic16_test_and_set(&num_sol_conf_alloc) > 1) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: trying to allocate the second instance of struct sol_config\n");
+		return NULL;
+	}
+	sol_conf = rte_calloc("sol_config", 1, sizeof(struct sol_config), 0);
+	INIT_LIST_HEAD(&sol_conf->req_queue.head);
+	return sol_conf;
+}
+
+int
+gk_solicitor_enqueue(struct sol_config *sol_conf, struct rte_mbuf *pkt,
+	uint8_t priority)
+{
+	struct priority_req *req_node = mb_alloc_entry(&sol_conf->mb);
```

Comment:
We have to test that `priority` is in [0, 63] because the code will break if this is false. If `priority` is not valid, the function should log a message and return an error.